src/Features/Core/Portable/ConvertAnonymousTypeToClass/AbstractConvertAnonymousTypeToClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,412 @@
+            var className = NameGenerator.GenerateUniqueName(
+                "NewClass", n => semanticModel.LookupSymbols(position, name: n).IsEmpty);
+
+            // First, create the set of properties this class will have based on the properties
+            // the anonymous type has itself.
+            var (properties, propertyMap) = GenerateProperties(anonymousType);
+
+            // Next, generate the full class that will be used to replace all instances of this
+            // anonymous type.
+            var namedTypeSymbol = await GenerateFinalNamedTypeAsync(
+                document, className, properties, cancellationToken).ConfigureAwait(false);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var editor = new SyntaxEditor(root, generator);
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var containingMember = anonymousObject.FirstAncestorOrSelf<SyntaxNode>(syntaxFacts.IsMethodLevelMember) ?? anonymousObject;
+
+            // Next, go and update any references to these anonymous type properties to match
+            // the new PascalCased name we've picked for the new properties that will go in
+            // the named type.
+            await ReplacePropertyReferencesAsync(
+                document, editor, containingMember, 
+                propertyMap, cancellationToken).ConfigureAwait(false);
+
+            // Next, go through and replace all matching anonymous types in this method with a call
+            // to construct the new named type we've generated.  
+            await ReplaceMatchingAnonymousTypesAsync(
+                document, editor, className,
+                containingMember, anonymousObject, 
+                anonymousType, cancellationToken).ConfigureAwait(false);
+
+            // Then, actually insert the new class in the appropriate container.
+            var container = anonymousObject.GetAncestor<TNamespaceDeclarationSyntax>() ?? root;
+            editor.ReplaceNode(container, (currentContainer, _) =>
+            {
+                var codeGenService = document.GetLanguageService<ICodeGenerationService>();
+                var options = new CodeGenerationOptions(
+                    generateMembers: true,
+                    sortMembers: false,
+                    autoInsertionLocation: false);
+
+                return codeGenService.AddNamedType(
+                    currentContainer, namedTypeSymbol, options, cancellationToken);
+            });
+
+            var updatedDocument = document.WithSyntaxRoot(editor.GetChangedRoot());
+
+            // Finally, format using the equals+getHashCode service so that our generated methods
+            // follow any special formatting rules specific to them.
+            var equalsAndGetHashCodeService = document.GetLanguageService<IGenerateEqualsAndGetHashCodeService>();
+            return await equalsAndGetHashCodeService.FormatDocumentAsync(
```

Comment:
>FormatDocumentAsync [](start = 53, length = 19)

I didn't understand this bit. Why are we formatting the entire `updatedDocument`? I'd expect we'd only format the generated code, but would leave pre-existing code alone.